### PR TITLE
Fix timeout regression that caused Node.js to not terminate

### PR DESCRIPTION
**Pull request by obartra**
Remaining timeout prevents node from terminating the process. These changes clear the timeout on promise resolution.

Closes https://github.com/sindresorhus/ky/issues/138

### CODE REVIEW

**Comment by szmarczak** [comment]
You need to mark `invalid timeout option` as failing and fix `prefixUrl option`.

**Comment by obartra** [comment]
I'm getting all tests to pass locally now but the `prefixUrl` seems to keep failing on CI. I don't see any more output, do you have suggestions as to what may be going on?

image

**Comment by sindresorhus** [comment]
I see more output in Travis. This is the failure details:

```
 /home/travis/build/sindresorhus/ky/node_modules/puppeteer/lib/Launcher.js:351
  Rejected promise returned by test. Reason:
  TimeoutError {
    message: 'Timed out after 30000 ms while trying to connect to Chrome! The only Chrome revision guaranteed to work is r669486',
  }
```

File: index.js [hunk 1]
```diff
@@ -115,27 +115,34 @@ class TimeoutError extends Error {
 	}
 }
 
-const delay = ms => new Promise((resolve, reject) => {
+function safeTimeout(resolve, ms, reject) {
```
Comment (szmarczak) [inline index.js]:
```suggestion
const safeTimeout = (resolve, reject, ms) => {
```

File: index.js [hunk 2]
```diff
@@ -115,27 +115,34 @@ class TimeoutError extends Error {
...
 		reject(new RangeError('The `timeout` option cannot be greater than 2147483647'));
-	} else {
-		setTimeout(resolve, ms);
 	}
-});
+
+	return setTimeout(resolve, ms);
+}
+
+const delay = ms => new Promise((resolve, reject) => safeTimeout(resolve, ms, reject));
```
Comment (szmarczak) [inline index.js]:
```suggestion
const delay = ms => new Promise((resolve, reject) => safeTimeout(resolve, reject, ms));
```

File: test/main.js [hunk 3]
```diff
@@ -186,17 +186,31 @@ test('timeout:false option', async t => {
...
-
 	const server = await createTestServer();
 	server.get('/', async (request, response) => {
-		requestCount++;
 		await delay(1000);
 		response.end(fixture);
 	});
 
 	await t.throwsAsync(ky(server.url, {timeout: 21474836470}).text(), RangeError, 'The `timeout` option cannot be greater than 2147483647');
-	t.is(requestCount, 1);
```
Comment (szmarczak) [inline test/main.js]: Why did you remove this? `requestCount` should be `0` here.

File: index.js [hunk 4]
```diff
@@ -50,7 +50,7 @@ const deepMerge = (...sources) => {
 					value = deepMerge(returnValue[key], value);
 				}
 
-				returnValue = {...returnValue, [key]: value};
+				returnValue = { ...returnValue, [key]: value };
```
Comment (szmarczak) [inline index.js]: Please don't make unrelated changes.

Comment (obartra) [reply index.js]: Yeah, for sure, I forgot to disable prettier, will update

File: index.js [hunk 5]
```diff
@@ -115,27 +115,38 @@ class TimeoutError extends Error {
...
 // `Promise.race()` workaround (#91)
-const timeout = (promise, ms, abortController) => new Promise((resolve, reject) => {
-	/* eslint-disable promise/prefer-await-to-then */
-	promise.then(resolve).catch(reject);
-	delay(ms).then(() => {
-		if (supportsAbortController) {
-			abortController.abort();
-		}
+const timeout = (promise, ms, abortController) =>
+	new Promise((resolve, reject) => {
```
Comment (szmarczak) [inline index.js]: It should be `=> new Promise`. No need to make a new line for that.

File: index.js [hunk 6]
```diff
@@ -115,27 +115,38 @@ class TimeoutError extends Error {
...
+
+		/* eslint-disable promise/prefer-await-to-then */
+		promise
+			.then(resolve)
+			.catch(reject)
+			.then(() => {
+				clearTimeout(timeoutID);
+			});
+		/* eslint-enable promise/prefer-await-to-then */
+	});
```
Comment (szmarczak) [inline index.js]:
```suggestion
});
```

File: index.js [hunk 7]
```diff
@@ -115,27 +115,38 @@ class TimeoutError extends Error {
...
+		}, reject, ms);
+
+		/* eslint-disable promise/prefer-await-to-then */
+		promise
+			.then(resolve)
+			.catch(reject)
+			.then(() => {
+				clearTimeout(timeoutID);
+			});
+		/* eslint-enable promise/prefer-await-to-then */
```
Comment (szmarczak) [inline index.js]: This looks much better:
```js
promise.then(resolve, reject).then(() => {
	clearTimeout(timeoutID);
});
```

File: index.js [hunk 8]
```diff
@@ -115,27 +115,38 @@ class TimeoutError extends Error {
 	}
 }
 
-const delay = ms => new Promise((resolve, reject) => {
+const safeTimeout = (resolve, reject, ms) => {
```
Comment (szmarczak) [inline index.js]: Detaching promises is not a good idea. I'd use `try`/`catch` here, but then this piece:

```js
try {
    safeTimeout(resolve, ms);
} catch (error) {
    reject(error);
}
```

would repeat twice. So to follow the DRY principle, I'm good with the current solution :+1:

**Comment by tiger-seo** [comment]
please, merge this, this bug is really annoying

**Comment by obartra** [comment]
Looks like I just needed to rebase, the built error went away now that it's up to date with master.

Regarding the style changes @szmarczak, would you be open to adding [prettier](https://prettier.io/) instead (in a separate PR)? This would remove the need for some of the current linting rules that cause build errors for stylistical / code consistency preferences.

**Comment by sholladay** [comment]
Use `xo --fix` if you want to fix the style issues automatically. It's significantly more opinionated than prettier. In a good way, if you ask me.

**Comment by obartra** [comment]
Oh nice I didn't realize that was an option, would you be open  to adding the`--fix` to the package.json script or as a precommit hook?

**Comment by sholladay** [comment]
We may accept an npm script. IIRC, Sindre has another project that has that, for discoverability.

But it's also very easy for us to take care of it before or after merging.

**Comment by tiger-seo** [comment]
hey guys, @obartra @szmarczak , please please merge this

**Comment by sindresorhus** [comment]
> Regarding the style changes @szmarczak, would you be open to adding prettier instead (in a separate PR)?

No thanks, I have developed a strong dislike for Prettier since I have to waste so much time on people accidentally submitting unwanted style changes because of Prettier.

> Oh nice I didn't realize that was an option, would you be open to adding the--fix to the package.json script or as a precommit hook?

No, you can just run `npx xo --fix`.
